File: 0061-rotate-list/0061-rotate-list.scala
object Solution:
    def rotateRight(head: ListNode, _k: Int): ListNode =
        var k = _k
        if k < 1 || head == null || head.next == null then
            head
        else
            var totalLen = 0
            var nthToLast: ListNode = null
            var last: ListNode = null

            def recur(node: ListNode): Int =
                if node == null then
                    k = k % totalLen
                    0
                else
                    totalLen += 1
                    if node.next == null then
                        last = node
                    val thisNum = 1 + recur(node.next)
                    if thisNum == k + 1 then
                        node.next = null
                    if thisNum == k then
                        nthToLast = node
                    thisNum
            recur(head)

            if k == 0 then
                head
            else
                last.next = head
                nthToLast

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = head
        if head is None:
            return head
        _next = curr.next
        len_list = 1
        tail = None

        while _next is not None:
            _next = _next.next
            len_list += 1
        if k > len_list:
            k = k % len_list
        num_nodes = len_list - k
        if num_nodes == 0:
            return head
        _next = curr.next
        
        while _next is not None:
            num_nodes -= 1 # this might need to be moved?
            if num_nodes == 0 and tail is None:
                tail = head
                head = _next
                curr.next = None
            curr = _next
            _next = curr.next
            if _next is None:
                curr.next = tail

        return head